surefire-api/src/main/java/org/apache/maven/surefire/booter/CommandReader.java

```diff
@@ -412,6 +412,10 @@ public void run()
                                 CommandReader.this.wakeupIterator();
                                 callListeners( command );
                                 break;
+                            case BYE_ACK:
+                                //After SHUTDOWN no more commands can come. Hence, do NOT go back to blocking in IO
+                                callListeners( command );
+                                return;
                             default:
                                 callListeners( command );
```

Comment:
The listeners are called here. But we can put IF condition:
`IF BYE_ACK -> return` at the end of the default case.

Reply:
Instead of calling the `return` we can make softer exit with `CommandReader.this.state.set( TERMINATED )`.

Reply:
Yes, I came to this same conclusion, change the state to TERMINATED.

Reply:
Changed.